presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewPlanValidator.java

```diff
@@ -77,29 +140,92 @@ protected Void visitLogicalBinaryExpression(LogicalBinaryExpression node, Materi
         return super.visitLogicalBinaryExpression(node, context);
     }
 
-    protected static final class MaterializedViewPlanValidatorContext
+    @Override
+    protected Void visitComparisonExpression(ComparisonExpression node, MaterializedViewPlanValidatorContext context)
+    {
+        if (!context.isProcessingJoinNode()) {
+            return super.visitComparisonExpression(node, context);
+        }
+
+        if (!node.getOperator().equals(ComparisonExpression.Operator.EQUAL)) {
+            throw new SemanticException(NOT_SUPPORTED, node, "Only EQUAL join is supported for materialized view.");
+        }
+
+        if (isValidationOnly) {
+            return super.visitComparisonExpression(node, context);
+        }
+
+        Field left = analysis.get().getScope(context.getTopJoinNode()).tryResolveField(node.getLeft())
+                .orElseThrow(() -> new SemanticException(
+                        NOT_SUPPORTED,
+                        node.getLeft(),
+                        "%s in join criteria is not supported for materialized view.", node.getLeft().getClass().getSimpleName()))
+                .getField();
+        Field right = analysis.get().getScope(context.getTopJoinNode()).tryResolveField(node.getRight())
+                .orElseThrow(() -> new SemanticException(
+                        NOT_SUPPORTED,
+                        node.getRight(),
+                        "%s in join criteria is not supported for materialized view.", node.getRight().getClass().getSimpleName()))
+                .getField();
+
+        Optional<TableColumn> leftBaseColumn = tryGetOriginalTableColumn(left);
+        Optional<TableColumn> rightBaseColumn = tryGetOriginalTableColumn(right);
+        if (leftBaseColumn.isPresent() && rightBaseColumn.isPresent() && !leftBaseColumn.get().equals(rightBaseColumn.get())) {
+            linkedBaseColumns.add(ImmutableList.of(leftBaseColumn.get(), rightBaseColumn.get()));
+        }
+
+        return super.visitComparisonExpression(node, context);
+    }
+
+    private Optional<TableColumn> tryGetOriginalTableColumn(Field field)
+    {
+        if (field.getOriginTable().isPresent() && field.getOriginColumnName().isPresent()) {
+            SchemaTableName table = toSchemaTableName(field.getOriginTable().get());
+            String column = field.getOriginColumnName().get();
+            return Optional.of(new TableColumn(table, column));
+        }
+        return Optional.empty();
+    }
+
+    public static final class MaterializedViewPlanValidatorContext
     {
-        private final Set<Join> joinNodes;
         private boolean isProcessingJoinNode;
+        private final LinkedList<Join> joinNodeStack;
```

Comment:
not a strong preference: At the moment, we support max one join node in the MV definition, then why do we need a stack?